staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -511,6 +511,13 @@ func (c *Cacher) Watch(ctx context.Context, key string, opts storage.ListOptions
 		return newErrWatcher(err), nil
 	}
 
+	// FIXME: Change the condition to ensure that we're not allocating memory for initEvents
+	//   without any reasonable boundary.
```

Comment:
what condition are you envisioning here?

Reply:
We sometime observe this problem that basically:
- kube-apiserver is starved (on network bandwidth is the most common example)
- the watches are being closed (because they aren't processing incoming events fast enough)
- the watches restart, but now all backlog lands in "initEvents"
- this continues and kube-apiserver memory is exploding (because the size of initEvents is bounded only on the size of transaction log in watchcache)

So what I was thinking about two things:
- ensuring that number of init_events stored in memory (across all watchers) is bounded
- [potentially] also introducing a QPS-based rate-limitting (though this one may cause other problems...)